src/app/data-model/DecodableList.h

```diff
@@ -293,21 +427,90 @@ class DecodableList : public detail::DecodableMaybeFabricScopedList<DataModel::I
         T mValue;
     };
 
-    // Need this->mReader and this->mFabricIndex for the name lookup to realize
-    // those can be found in our superclasses.
+private:
+    inline const TLV::TLVReader & reader() const
+    {
+        VerifyOrDie(readerOrSpan.mIsDecodable);
+        return readerOrSpan.mValue.mHolder.mReader;
+    }
+
     template <typename T0 = T, std::enable_if_t<DataModel::IsFabricScoped<T0>::value, bool> = true>
-    Iterator begin() const
+    inline void FabricInit()
     {
-        return Iterator(this->mReader, this->mFabricIndex);
+        readerOrSpan.mValue.mHolder.mFabricIndex.ClearValue();
     }
 
     template <typename T0 = T, std::enable_if_t<!DataModel::IsFabricScoped<T0>::value, bool> = true>
-    Iterator begin() const
+    inline void FabricInit()
+    {}
+
+    // A container of bool + either a DataModel::List or a mReader for DecodableList
+    class ValueHolder
     {
-        return Iterator(this->mReader);
-    }
+    public:
+        inline ValueHolder() {}
+        template <typename... Params>
+        inline ValueHolder(Params &&... params) : mValue(std::forward<Params>(params)...)
+        {}
+
+        union Value
+        {
+            inline Value() {}
+            template <typename... Params>
+            inline Value(Params &&... params) : mList(std::forward<Params>(params)...)
+            {}
+
+            detail::DecodableHolder<DataModel::IsFabricScoped<T>::value> mHolder;
+            DataModel::List<const T> mList;
+        } mValue;
+        bool mIsDecodable = false;
+    };
+
+    ValueHolder readerOrSpan;
 };
 
+template <typename X>
+inline CHIP_ERROR Encode(TLV::TLVWriter & writer, TLV::Tag tag, DecodableList<X> list)
+{
+    TLV::TLVType type;
+
+    ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Array, type));
+    CHIP_ERROR result = list.IterateGeneric([&writer](const auto & element, bool &) -> CHIP_ERROR {
+        ReturnErrorOnFailure(Encode(writer, TLV::AnonymousTag(), element));
+        return CHIP_NO_ERROR;
+    });
+    ReturnErrorOnFailure(result);
+    return writer.EndContainer(type);
+}
+
+template <typename X, std::enable_if_t<DataModel::IsFabricScoped<X>::value, bool> = true>
+inline CHIP_ERROR EncodeForWrite(TLV::TLVWriter & writer, TLV::Tag tag, DecodableList<X> list)
+{
+    TLV::TLVType type;
+
+    ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Array, type));
+    CHIP_ERROR result = list.IterateGeneric([&writer](const auto & element, bool &) -> CHIP_ERROR {
+        ReturnErrorOnFailure(EncodeForWrite(writer, TLV::AnonymousTag(), element));
+        return CHIP_NO_ERROR;
+    });
+    ReturnErrorOnFailure(result);
+    return writer.EndContainer(type);
+}
+
+template <typename X, std::enable_if_t<DataModel::IsFabricScoped<X>::value, bool> = true>
+inline CHIP_ERROR EncodeForRead(TLV::TLVWriter & writer, TLV::Tag tag, FabricIndex accessingFabricIndex, DecodableList<X> list)
```

Comment:
This should be checking that `accessingFabricIndex` matches `list`, no?

Reply:
🤷‍♂️ `List<T>` doesn't - should it? (we may as well address both in this PR, as it's just a clean-up PR now)